Add support for concurrent send/recv apis for tun device

kp-mariappan-ramasamy opened the pull request:
Tokio supports async api's for UDP socket which does not need `&mut self`

    pub async fn recv(&self, buf: &mut [u8]) -> Result<usize>
    pub async fn send(&self, buf: &[u8]) -> Result<usize>

 This makes it pretty easy to call it concurrently, without adding Arc<Mutex<T>> to UdpSocket.

This PR tries to add support for similar apis to tun device in `tun2` crate.

Reference:
    https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#method.recv
    https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#method.send

This PR does not take care of windows, which is not as straight-forward as posix variants.

kp-mariappan-ramasamy committed:
Add send and recv apis similar to Tokio::UdpSocket

Tokio supports async api's for UDP socket which does not need `&mut self`

pub async fn recv(&self, buf: &mut [u8]) -> Result<usize>
pub async fn send(&self, buf: &[u8]) -> Result<usize>

This makes it pretty easy to call it concurrently, without adding Arc<Mutex<T>> to
UdpSocket.

This PR tries to add support for similar apis to tun device.
In this commit, only sync version of recv/send apis are added. Async version will be added in
further commit.

Reference:
https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#method.recv
https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#method.send

kp-mariappan-ramasamy committed:
Add async unix send/recv apis for unix

Since unix uses AsyncFd from tokio, it is easier to wait for Ready
on AsyncFd and call sync send/recv apis

This commit does not take care of windows, which is not as straight-forward as posix variants.

xmh0511 commented on src/platform/posix/split.rs:
```diff
@@ -68,12 +68,39 @@ pub struct Reader {
     }
+
+    pub(crate) fn recv(&self, mut buf: &mut [u8]) -> io::Result<usize> {
+        let mut stack_buf = [0u8; crate::DEFAULT_MTU as usize + PIL];
```
Based on we want to do some optimization here, let us move the `stack_buf` to line 90 here to avoid each useless initialization of `stack_buf` every time we enter `recv` without using `stack_buf`. The same is true for `send`.

kp-mariappan-ramasamy replied:
Sounds good. Updated

xmh0511 commented on src/platform/posix/split.rs:
```diff
@@ -106,12 +133,43 @@ pub struct Writer {
     }
+
+    pub(crate) fn send(&self, buf: &[u8]) -> io::Result<usize> {
+        let mut stack_buf = [0u8; crate::DEFAULT_MTU as usize + PIL];
```
See above. We still can compare `crate::DEFAULT_MTU as usize + PIL` and `len` first, then determine whether we need `stack_buf`.

xmh0511 replied:
![QQ_1721700667141](https://github.com/user-attachments/assets/41cbab0b-5481-4b38-8bcd-eda4174567e5)
![QQ_1721701049208](https://github.com/user-attachments/assets/2e96ba1b-4174-46da-a036-681d83a6ef2e)

xmh0511 commented:
BTW, if we don't plan to implement `recev` and `send` on the Windows platform in a short time, we should leave the implementation of them on Windows with `unimplemented!()` to avoid making the platform difference.

kp-mariappan-ramasamy committed:
Refactor to use STACK_BUF_LEN as const

kp-mariappan-ramasamy commented:
Thanks @xmh0511 for the review. I have addressed the comments and tested in linux and android platform.
Also added unimplemented!() api for windows platform. 

Can you please take a look again ?

xmh0511 commented:
> Thanks @xmh0511 for the review. I have addressed the comments and tested in linux and android platform. Also added unimplemented!() api for windows platform.
> 
> Can you please take a look again ?

Everything sounds good to me, I will merge this PR after checking passed.
